### PR TITLE
[HOLD] Yaml parser

### PR DESCRIPTION
**Pull request by aileen**
Nothing to see here 🙈 

----
Todos:
- [x] make code async
- [ ] what is the expected behaviour if default routes file is missing (super edge case)?
- [x] write tests
- [ ] get feedback on behaviour and naming
- [ ] test with Ghost CLI PR TryGhost/Ghost-CLI#681
For later:
- [ ] finalise this PR as soon as dynamic routing is PRd
- [ ] expand tests to parse more complicated YAML files

### CODE REVIEW

**Comment by AileenCGN** [comment]
@kirrg001 / @ErisDS if anyone wants to have a first look at this?

Two questions, that are still open for me:

1. what is the expected behaviour if default routes file is missing (super edge case)?
2. is the behaviour and naming ok from your POV?

**Comment by kirrg001** [comment]
> what is the expected behaviour if default routes file is missing

How is that possible? 🤔 Doesn't we keep a copy of it in core?

**Comment by AileenCGN** [comment]
> How is that possible? 🤔 Doesn't we keep a copy of it in core?

Yes, we do. That's why I said 'super edge case'. Might just be the case of someone deleting it from core. Or can you think of any other reason? 🤔

File: content/settings/README.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# Content / Settings
+
+### routes.yaml
+
+This is how the default `routes.yaml` file looks like:
+
```
Comment (AileenCGN) [inline content/settings/README.md]: I was thinking of adding some more text here with a link to the docs as soon as those are in place.

File: core/server/config/overrides.json [hunk 2]
```diff
@@ -6,6 +6,7 @@
         "helperTemplates": "core/server/helpers/tpl/",
         "adminViews": "core/server/web/admin/views/",
         "defaultViews": "core/server/views/",
+        "defaultRoutes": "core/server/services/route/",
```
Comment (AileenCGN) [inline core/server/config/overrides.json]: Not sure, if this is needed, and we'll still change the location for the default routes, but for now I thought it makes sense.

File: core/server/services/route/settings.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+                return fs.readFile(routesFile, 'utf8');
+            }).then((file) => {
+                debug('Default routes file copied');
+                return Promise.resolve(file);
+            }).catch((err) => {
+                // the default routes file doesn't exist 😱, or we can't access
+                // the content/settings folder. Now we have the salad
+                // TODO: what error do we want to return here
+                return Promise.reject(new common.errors.GhostError({
+                    message: 'Error trying to copy the default routes file.',
```
Comment (AileenCGN) [inline core/server/services/route/settings.js]: The only other errors I can actually think of are 
- error with usage of a storage adapter
- missing `content/settings` directory (user deleted it, or it's not there yet)
- no permission to access the `content/settings` directory (CLI case usually)

Not sure, if you can think of any more cases...

File: core/server/services/route/settings.js [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+                err: error,
+                context: error.path
+            }));
+        });
+}
+
+loadSettings = function loadSettings() {
+    return ensureRoutesFile().then((yamlFile) => {
+        try {
+            const routes = yaml.safeLoad(yamlFile);
```
Comment (AileenCGN) [inline core/server/services/route/settings.js]: wrt `js-yaml`, I am not sure if it would be better to `Promisify` it?

**Comment by kirrg001** [comment]
> Yes, we do. That's why I said 'super edge case'. Might just be the case of someone deleting it from core. Or can you think of any other reason? 

You can ignore this case :)

**Comment by ErisDS** [comment]
I have written the spec into an issue that is here: https://github.com/TryGhost/Ghost/issues/9528 IMO this PR mixes up 2 things. 1) the concept of YAML settings and 2) the first potential settings file.

What we need is a generic system for settings files first. Adding actual file IMO comes later.

**Comment by AileenCGN** [comment]
> What we need is a generic system for settings files first. Adding actual files IMO comes later.

Sorry, this wasn't clear to me _at all_ from the information I had available at the time when I started. Maybe I misunderstood it. Reading the spec now, makes it a lot clearer to me.

**Comment by AileenCGN** [comment]
I created a new PR #9534 which is only the parser and reads the `content/settings` directory.

**Comment by AileenCGN** [comment]
Closing this in favour of #9534
